### ISSUE TITLE
Difference between implementation and equation (1)

### ISSUE THREAD

**Issue by skoohy** [thuml/Transolver#10]
In the paper, it is written that 

![Image](https://github.com/user-attachments/assets/b3e64d88-bf54-4785-b21a-bb7a2d3bd755)

where Project() is either a point-wise linear layer or a local convolution. In the code, specifically in `PDE-Solving-StandardBenchmark/model/Physics_Attention.py`, there is two separate convolutions `self.in_porject_x` and `self.in_project_fx` which are used to compute `x_mid` and `fx_mid`, respectively. I'm not totally sure what their purpose is. Since Project() projects C channels into M weights, I would expect that `self.in_project_slice` represents Project(). From the equation, it would look like we would do something like 

```
x = x.transpose(0, 2, 3, 1).reshape(B, N, self.num_heads, self.head_dim).transpose(0, 2, 1, 3)

slice_weights = nn.softmax(
            self.in_project_slice(x)
            / jnp.clip(self.temperature, min=0.1, max=5),
            axis=-1,
        )

slice_norm = slice_weights.sum(2)
slice_token = jnp.einsum("bhnc, bhng -> bhgc", x, slice_weights)
````
But, in the code we do  
```
fx_mid = (
            self.in_project_fx(x)
            .transpose(0, 2, 3, 1)
            .reshape(B, N, self.num_heads, self.head_dim)
            .transpose(0, 2, 1, 3)
        )  
        
x_mid = (
           self.in_project_x(x)
           .transpose(0, 2, 3, 1)
           .reshape(B, N, self.num_heads, self.head_dim)
           .transpose(0, 2, 1, 3)
        )  

slice_weights = nn.softmax(
            self.in_project_slice(x_mid)
            / jnp.clip(self.temperature, min=0.1, max=5),
            axis=-1,
        )

slice_norm = slice_weights.sum(2)
slice_token = jnp.einsum("bhnc, bhng -> bhgc", fx_mid, slice_weights)
```
I'm not sure why we are using or need `x_mid` or `fx_mid` and why we would use `x_mid` for the `nn.softmax` and then `fx_mid` for the `slice_token`. The second line of equation (1) also shows that the input to Project() is what is used to compute $s_j$ but in the code `fx_mid` is used instead of `x`? The code I provided is written slightly differently since I reimplemented it in jax, but it's pretty much the same. Maybe I missed where this is mentioned in the paper?

As a very small test, I ran the code without the convolutions for 10 epochs and got an L2 error of 0.45 on the test set, and for the original code an L2 error of 0.03. So at least up to that point, they seem to have an important purpose.

Any help or clarification would be greatly appreciated, thanks so much.

**Comment by wuhaixu2016** [comment]
Hi, thanks for your detailed questions and experiments.

The original implementation in Transolver follows the style of self-attention mechanism, where the calculation of attention weights and values are based on different projection layers. Thus, we adopt two linear layers to yield slice weights (in_project_x) and aggregate information (in_project_fx) respectively.

However, in our recent study (Transolver++, an upgraded version of Transolver, Algorithm 1 in https://arxiv.org/pdf/2502.02414v1 ), we find this implementation redundant. Thus, in Transolver++, we have removed `fx_mid` and rewritten the slice process as `slice_token = jnp.einsum("bhnc, bhng -> bhgc", x_mid, slice_weights)`. Note here we use **x_mid** to replace fx_mid, not x. This design can further boost the single-GPU input capacity to 1 million mesh points and bring better results in most cases.

As for why x_mid is better than x, I can only provide some intuitive understanding. Specifically, you can understand this from the self-attention design: why we need to adopt an additional linear layer for values (v), not just adopt the input (x). My opinion is that: In Transformer-style architectures, the input (x) will be re-used in the next layer as residual. Thus If we directly use input without any buffer layer, this design will cause an unnecessary trade-off problem for x that one single representation needs to be residual and be values for attention at the same time, thereby damaging the final performance.

Also, for your small experiments, I truly appreciate that. However, 10-epochs is far from the 500 epochs in our official experiments. Thus, I think the comparison between 10-epochs and 500-epochs may be meaningless.

**Comment by skoohy** [comment]
Thank you for your response @wuhaixu2016 ! 

So you're saying that using x for both the input into the self-attention and the residual condition implies that there will be some sort of "trade-off" in using x for both? Instead, use x as the residual connection as usual, and similar to how the standard self-attention works nothing within the mechanism directly uses x by itself but, instead we do some embedding of x beforehand. Here we do the same thing by using some embedding before going through the Physics attention. I'm assuming one might also be able to use say linear layer here instead of convolutions but the convolution may be better for encoding local information?

Also, are there plans to release the code for Transolver++ in the future?

Thank you!
